docs/performance-guide.md

```diff
@@ -25,150 +54,207 @@ guide intends to answer the following questions:
+pattern etc.) in the limited paragraph, however, it provides some evaluation methods to help users to approximately evaluate their requirement of the inbound or outbound message, then find the proper tiers by checking the performance table. 
 
-Different user has different use cases. It is impossible for this document to enumerate all
-scenarios (different use case, different message size, or message sending
-pattern etc.) in the limited paragraph, however, it provides a formula for users to approximately
-evaluate their specific max inbound or outbound based on the 4 known use cases.
+<a name="perf_insight"></a>
+Performance insight
+-------------------
 
-Guide
------
+This section first lists all factors that take impact on performance, then provides methods to help evaluate the performance requirements.
+
+<a name="perf_factors"></a>
+Performance factors
+-------------------
+
+Theoretically, ASRS' capacity is limited by computation resources: CPU, Memory, and network. For example, more connections to ASRS, more memory ASRS consumed. For larger message traffic, for example, every message is larger than 2048 bytes, it requires ASRS to spend more CPU cycles to process as well. Meanwhile, Azure network bandwidth also imposes a limit for maximum traffic.
 
-The inbound and outbound capacity is impacted by the following factors:
+The transport type, Websockets, Sever-Sent-Event, or Long-Polling is another factor which impacts performance. For the same API, Websockets shows the best performance, Sever-Sent-Event is worse, and Long-Polling is the worst. ASRS recommends Websockets by default.
 
--   unit tier (CPU/Memory quota)
+In addition, the cost of message routing also limits the performance. ASRS plays a role of router, which routes the message from a bunch of clients or servers to other clients or servers. Different scenario or API requires different routing. For **echo**, the client sends a message to itself, and the routing destination is also itself. This is the lowest cost routing. But for **broadcast**, **send to group**, **send to connection**, ASRS has to find the target connections through the internal distributed data structure, which consumes more CPU, Memory and even network bandwidth. As a result, performance is not as good as **echo**.
+
+On the default mode, the app server may also become a bottleneck for some scenarios, because Azure SignalR SDK has to invoke the Hub, meanwhile it maintains the connection with every client connection through heart-beat.
+
+On server-less mode, the client sends message by HTTP post which is not as efficient as Websockets.
+
+As a summary, the inbound and outbound capacity is impacted by the following factors:
+
+-   SKU tier (CPU/Memory)
 
 -   connection number
 
 -   message size
 
--   message send/receive rate
+-   message send rate
+
+-   transport type
```

Comment:
Do we have a comparison between JSON and messagepack?

Reply:
This has been done before, but not yet on nightly perf track. I can setup a nightly track for Echo + Messagepack if you think it is necessary.

Reply:
A quick test indicates app server's CPU usage can be reduced. Let me collect this information, and update the doc.

Reply:
ok, then here "transport type" means websockets or long-polling right?

Reply:
Yes. json/messagepack is protocol.

Reply:
The Echo + Msgpack data indicates that on app server side, the CPU usage reduction is not as apparent as expected. I think I need to collect more data. I prefer to mention this in doc but not highlight it because it does not impact the performance on ASRS side.

Average CPU usage on app servers:

Appserver CPU | JSON (Avg CPU) | MsgPack (Avg   CPU)
-- | -- | --
1 | 22%~29% | 25%~29%
2 | 43.48%~48.12% | 38%~50%
5 | 89.78%~104.6% | 86%~90%
10 | 119%~148.19% | 131%~135%
20 | 164%~202.76% | 155%~187%
50 | 162.17%~189.68% | 149%~203%
100 | 175%~246% | 186%~247%

Max CPU usagge on app servers:

Appserver CPU | JSON (Max CPU) | MsgPack (Max   CPU)
-- | -- | --
1 | 33.3%~46.7% | 46.70%
2 | 53.3%~62.5% | 60%~80%
5 | 126.7%~153.3% | 133%~144%
10 | 173.3%~220% | 175%
20 | 233.3%~253.3% | 200%~220%
50 | 186.7%~220% | 180%~240%
100 | 206%~280% | 213%~280%